File: backend/src/main/kotlin/com/nevexis/backend/schoolManagement/requests/Request.kt
@file:UseSerializers(
    BigDecimalSerializer::class
)

package com.nevexis.backend.schoolManagement.requests


import com.nevexis.backend.schoolManagement.users.OneRoleUser
import com.nevexis.backend.schoolManagement.users.User
import com.nevexis.backend.schoolManagement.users.UserView
import com.nevexis.backend.serializers.BigDecimalSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.json.JsonContentPolymorphicSerializer
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import java.math.BigDecimal
import java.time.LocalDateTime

data class Request(
    val id: Int,
    val requestedByUser: UserView,
    val requestValue: RequestValue,
    val requestDate: LocalDateTime,
    val requestStatus: RequestStatus = RequestStatus.PENDING,
    val resolvedByUser: UserView? = null,
    val resolvedDate: LocalDateTime? = null
)

data class AdditionalRequestInformation(
    val valueId: BigDecimal,
    val status: RequestStatus? = null,
    val schoolId: BigDecimal,
    val periodId: BigDecimal
)

sealed class RequestValue {
    data class UserRegistration(
        val user: User,
        val status: RequestStatus? = null
    ) : RequestValue()

    data class Role(
        val oneRoleUser: OneRoleUser,
        val status: RequestStatus? = null
    ) : RequestValue()
}

@Serializable(with = RequestValueSerializer::class)
sealed class RequestValueJson {
    @Serializable
    data class UserRegistration(
        val schoolUserPeriodId: Int,
        val status: RequestStatus? = null
    ) : RequestValueJson()

    @Serializable
    data class Role(
        val schoolUserRolePeriodId: Int,
        val status: RequestStatus? = null
    ) : RequestValueJson()
}

object RequestValueSerializer :
    JsonContentPolymorphicSerializer<RequestValueJson>(RequestValueJson::class) {
    override fun selectDeserializer(element: JsonElement) = when {
        "schoolUserPeriodId" in element.jsonObject -> RequestValueJson.UserRegistration.serializer()
        "schoolUserRolePeriodId" in element.jsonObject -> RequestValueJson.Role.serializer()
        else -> error("There are no other types of requests")
    }
}

enum class RequestStatus {
    PENDING,
    REJECTED,
    APPROVED
}

File: frontend/src/model/Request.ts
import {User, UserView} from "./User";
import {RequestStatus} from "./RequestStatus";

export type RequestValue = UserRegistration | Role

export class Request {
    id: number
    requestedByUser: UserView
    requestValue: RequestValue
    requestDate: Date
    requestStatus: RequestStatus
    resolvedByUser: UserView | null
    resolvedDate: Date | null

    constructor(id, requestedByUser, requestValue, requestDate, requestStatus, resolvedByUser, resolvedDate) {
        this.id = id
        this.requestedByUser = requestedByUser
        this.requestValue = requestValue
        this.requestDate = requestDate
        this.requestStatus = requestStatus
        this.resolvedByUser = resolvedByUser
        this.resolvedDate = resolvedDate
    }

}

export class UserRegistration {
    user: User
    status: RequestStatus | null

    constructor(user, status) {
        this.user = user
        this.status = status
    }
}

export class Role {
    oneRoleUser: User
    status: RequestStatus | null

    constructor(oneRoleUser, status) {
        this.oneRoleUser = oneRoleUser
        this.status = status
    }
}
